drivers: nrf: Fix UART TX hanging when using both polling and async API.

Mierunski opened the pull request:
Fix UART TX hanging when using both polling and async API.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

fixes #22264

Mierunski committed:
drivers: nrf: Fix UART TX hanging when using both polling and async API.

Add locking between async and polling tx transfers.
Add safety counters while waiting for transmission to finish.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

JordanYates commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -229,6 +228,11 @@ static void uart_nrfx_poll_out(struct device *dev,
 	 * blue moon against hanging up the whole thread permanently
 	 */
 
+#ifdef CONFIG_UART_0_ASYNC
```
I don't think this fixes the core problem, which is that the condition listed in the comment is not actually applied.
'The UART API dictates that poll_out should wait for the transmitter to be empty before sending a character'
Given that UART doesn't have the TXSTARTED/TXSTOPPED events, perhaps polling until the TXDRDY interrupt is not enabled would work?

Mierunski replied:
Problem here was that event was cleared in ISR instead of poll_out function. Condition in comment is applied and working.

nordic-krch replied:
why not use same mechanism for business detection as in async call?
`if (atomic_cas(&txbuf, NULL, &c) == false) return;` to just skip poll out when async transfer is ongoing and make driver busy for that time. 

Evantually, we will convert to tty (console and logger) so dropping bytes on conflict should be accepted.

JordanYates commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -229,6 +228,11 @@ static void uart_nrfx_poll_out(struct device *dev,
 
+#ifdef CONFIG_UART_0_ASYNC
+	/* Disable TXDRDY interrupts, so that ISR doesn't clear TXDRDY event. */
+	nrf_uart_int_disable(uart0_addr, NRF_UART_INT_MASK_TXDRDY);
```
Disabling the interrupt will stop the thread locking up, at the expense of dropping any data remaining in the ASYNC send. I also don't think this fixes two threads calling poll_out at the same time, they just both disable the interrupt and still both loop on TXDRDY at the same time.

Mierunski replied:
Function is working fine with two threads calling poll_out at the same time, it is possible to lose some bytes but there is no other solution if we want to keep it lockless.
If writing from multiple contexts is desired, tty api should be used.
Not all data is dropped, it depends on timings but most of the data is sent out from async send.

JordanYates replied:
As far as I can tell, there is no way to force the logger backend or printk to route through the TTY system. If this is true, there is no way to reliably send data via UART, as printk/logger can blurt over the tty'd TX at any point in time. I understand the desire to keep it lockless, but it seems to come at the cost of making UART unusable for data protocols, rather than purely debug output.

JordanYates replied:
This doesn't seem to be the case for the other UART drivers I have scanned through, which have a dedicated TXBUSY register or similar.  If a lock is required to have this driver conform to the majority, that seems a worthy trade to me.

Mierunski committed:
drivers: nrf: Fix UARTE TX hangs when using both polling and async API.

Added locking for TX transfers between async and polling API.
Added safety counters for checking if transmission finished.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

JordanYates commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -485,6 +483,12 @@ static int uarte_nrfx_tx(struct device *dev, const u8_t *buf, size_t len,
 	data->async->tx_buf = buf;
```
There is still a chance of incorrect behavior here. 
If this function is pre-empted after the check above, but before these 2 lines run, the check is no longer valid.
Thread 2 could have started a TX, this function returns and overwrites data->async.
The original TX will never have the user callback called on.

nordic-krch replied:
yep, for that i would recommend `atomic_cas` something like:
`if (atomic_cas(&txbuf, NULL, buf)==false) return -EBUSY;`

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -242,10 +246,18 @@ static void uart_nrfx_poll_out(struct device *dev,
+		/* Deactivate the transmitter so that it does not needlessly
+		 * consume power.
+		 */
+		nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STOPTX);
```
Interrupt driven mode also needs the transmitter to stay active [as long as the TX interrupt is enabled](https://github.com/zephyrproject-rtos/zephyr/pull/22355/files#diff-24c27a01c8713b6b093dfd80f40883d0L745), so it seems that you should also use `device_busy_check(dev)` here (or introduce some flag) to handle that case.

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -915,33 +931,31 @@ static void uarte_nrfx_poll_out(struct device *dev, unsigned char c)
-				      NRF_UARTE_INT_ENDTX_MASK |
-				      NRF_UARTE_INT_TXSTOPPED_MASK);
-	}
+		lock = &data->async->tx_size;
```
There is a chance that once transfer will be completed a new one will come. In that case, if in interrupt we overwrite length (*lock = 1).  Maybe we should lock immediately in while loop which is doing the check? Something like:
```
do {
  while(txbuf) {  uarte_nrfx_isr_async(dev); }
} while (atomic_cas(&lock, 0, 1) == false);
```

Mierunski replied:
If a new one comes it will be handled normally, in worst case scenario we will lose first byte send by poll_out

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -915,33 +931,31 @@ static void uarte_nrfx_poll_out(struct device *dev, unsigned char c)
+			 */
+			k_sleep(1);
+			if (--safety_cnt == 0) {
+				break;
```
I guess it would be safer to return than to continue without obtaining the lock.

Mierunski replied:
We end up here if current thread is using poll out while lower priority thread is in middle of transaction. If we return here then successive poll outs from higher priority thread would be dropped, if we continue then we can lose first byte from higher priority thread. This point actually should never be reached as k_sleep will allow lower priority threads to execute and finish their operations. Safety counter was added just in case

Mierunski commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -209,26 +209,54 @@ static int uart_nrfx_poll_in(struct device *dev, unsigned char *c)
 	 */
+	lock = &uart0_cb.tx_buffer_length;
+#else
+	static atomic_val_t poll_out_lock;
```
There is only single instance of UART. In UARTE there are locks per instance

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -239,13 +267,17 @@ static void uart_nrfx_poll_out(struct device *dev,
 
-	/* Deactivate the transmitter so that it does not needlessly consume
-	 * power.
+	NRFX_WAIT_FOR(event_txdrdy_check(), 1000, 1, res);
```
I don't entirely get the idea behind this change. When can it happen that the transmission of the character does not finish?

Mierunski replied:
There is very small risk that if asynchronous TX started in line 937 during poll out in interrupt and ended in lines 964-965 then asynchronous call will clear ENDTX event and result in hanging that poll_out in interrupt.

Mierunski commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -209,26 +209,54 @@ static int uart_nrfx_poll_in(struct device *dev, unsigned char *c)
+			}
+		}
+	} else {
+		*lock = 1;
```
Lock here is written so that uart_tx returns busy if called from interrupt context.
If another interrupt comes then it is possible that we will lose first character but further calls to poll_out will result in successful transmission

Mierunski commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -209,26 +209,54 @@ static int uart_nrfx_poll_in(struct device *dev, unsigned char *c)
-	 * blue moon against hanging up the whole thread permanently
+	atomic_t *lock;
+#ifdef CONFIG_UART_0_ASYNC
+	while (uart0_cb.tx_buffer) {
```
fixed

Mierunski commented:
@JordanYates Please check if your issue is fixed

JordanYates commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -479,12 +478,20 @@ static int uarte_nrfx_tx(struct device *dev, const u8_t *buf, size_t len,
 	}
 
-	if (data->async->tx_buf || data->async->tx_size) {
+	if (atomic_cas((atomic_t *) &data->async->tx_size,
```
The atomic compare and set here is on async->tx_size , but the busy check in poll_out is on ansyc->tx_buf , so the pre-emption problem still exists.
1. Compare and set passes
2. Pre-empted
3. poll_out is called
4. Both threads think they are clear to use the hardware
I think you could just swap the check out to be on tx_buf, but I'm finding it hard to reason about the other ways tx_size is used.

Mierunski replied:
It has been taken into account, poll_out also has lock on async->tx_size and it will be blocked there in described scenario. Reason for using both tx_buf and tx_size for locks is that tx_buf is set only with async transfers, this way if poll_out is called from interrupt context then async transfer can be finished when trying to lock tx_buf, while tx_size is locked only in thread context, so if during poll_out its pre-empted and poll_out is called then we can still send from interrupt context. Only risk is that first byte sent using poll_out from interrupt context can be lost.
As using blocking functions in interrupt context is bad practice I assume that it would most likely occur only in hardfault handler, we don't want to lose that information so we cannot simply ignore poll out.
Current solution allows for using both async and polling API from different contexts without queueing output at cost of possible losing single byte when switching to interrupt context.

JordanYates replied:
Is that the only time bytes should be lost now? I have observed the first character of my test print disappearing with the setup described in my issue.

Mierunski replied:
Note that logger uses some multi byte encodings, depending on terminal settings it may seem like some characters are lost when logger thread is preempted in multi byte enconding. I noticed it when using ansi encoding in terminal, but looking at raw data I haven't seen it, therefore its not a UART issue.

JordanYates replied:
As you say, the raw bytes are there. Issue appears when we preempt after the first byte of the color code escape sequence. Thanks for the explanation, everything looks good to me then.

JordanYates commented:
I can confirm that the core problem is fixed, thanks @Mierunski 
There is just my comment on my pre-emptibility to resolve, and I am happy.

Mierunski commented:
@ioannisg ping
